lib/rules/prefer-object-has-own.js

```diff
@@ -0,0 +1,90 @@
+/**
+ * @fileoverview Prefers Object.hasOwn instead of Object.prototype.hasOwnProperty
+ * @author Nitin Kumar
+ * @author Gautam Arora
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+const astUtils = require("./utils/ast-utils");
+
+/**
+ * Checks to see if a property name object exists in the subtree recursively.
+ * @param {ASTnode} node to evalutate.
+ * @returns {boolean} `true` if object property exists, `false` otherwise.
+ */
+function hasLeftHandObject(node) {
+    if (!node.object) {
+        return false;
+    }
+
+    /*
+     * Object.hasOwnProperty.call(obj, prop) or ({}).hasOwnProperty.call(obj, prop) - `true`
+     * ({ foo }.hasOwnProperty.call(obj, prop)) - `false`, object literal should be empty
+     */
+    if (node.object.name === "Object" || (node.object.type === "ObjectExpression" && node.object.properties.length === 0)) {
+        return true;
+    }
+
+    const propertyName = astUtils.getStaticPropertyName(node.object);
+
+    if (propertyName === "hasOwnProperty" || propertyName === "prototype") {
+        return hasLeftHandObject(node.object);
+    }
```

Comment:
These would be false positives:

```js
({}).prototype.hasOwnProperty.call(a, b);

Object.prototype.prototype.hasOwnProperty.call(a, b);

Object.hasOwnProperty.prototype.hasOwnProperty.call(a, b);
```

Do we need recursion here, since this rule targets only 3 exact patterns?

Reply:
I have updated the code to not to use recursion and added these test cases.